### PR TITLE
[Issue-2192] Improve attestation prioritization for block proposal

### PR DESCRIPTION
**Pull request by mbaxter**
## PR Description
Reprioritize attestations as on-chain attestations are processed.  Previously, we prioritized attestations based on total number of set attester bits.  With these changes, we now prioritize by number of attester bits that have not been seen on chain in another block.  This should make it easier to collect high value attestations for inclusion in proposed blocks.

## Fixed Issue(s)
Part of #2192

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/MatchingDataAttestationGroup.java [hunk 1]
```diff
@@ -119,32 +122,91 @@ public long size() {
...
+              final Set<ValidateableAttestation> attestationsByOldCount =
+                  attestationsByValidatorCount.getOrDefault(
+                      oldValidatorCount, Collections.emptySet());
+
+              // Remove attestations and re-add so that the priority is recalculated
+              attestationsByOldCount.removeAll(attestations);
+              attestations.forEach(this::add);
+
+              // We're processing in reverse order so items should only move to lower values
+              // So, should we safe to remove here
```
Comment (cemozerr) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/MatchingDataAttestationGroup.java]:
```suggestion
              // So, should be safe to remove here
```

File: ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/attestation/MatchingDataAttestationGroupTest.java [hunk 2]
```diff
@@ -153,6 +153,54 @@ public void iterator_shouldAggregateAttestationsWithMoreValidatorsFirst() {
...
+    final ValidateableAttestation attestationC = addAttestation(4, 2, 6, 10, 14, 18);
+    final ValidateableAttestation attestationD = addAttestation(0, 3, 7);
+
+    // Remove some attesters, which should reprioritize the attestations
+    group.remove(toRemove);
+
+    final ValidateableAttestation attestationBC =
+        ValidateableAttestation.fromAttestation(
+            aggregateAttestations(attestationB.getAttestation(), attestationC.getAttestation()));
+    assertThat(group).containsExactly(attestationBC, attestationA, attestationD);
```
Comment (cemozerr) [inline ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/attestation/MatchingDataAttestationGroupTest.java]: I'm not sure if I understand this test case. The aggregate of attestation B and C seems to be the one with a higher validator count regardless of attestation prioritization. What am I missing?

**Review by cemozerr — APPROVED** [review]
Had one question about a test case. Rest looks good to me.

File: ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/attestation/MatchingDataAttestationGroupTest.java (hunk 2 above)
Comment (mbaxter) [reply ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/attestation/MatchingDataAttestationGroupTest.java]: Good catch.  The main point was to just do a more complicated update, but you're right it doesn't make any difference to the final sorting.  Updated.

**Comment by ajsutton** [comment]
As I understand it, this is reprioritising the attestations within a matching data group - we already do our best to aggregate those so I'm not sure how many attestations with the same data wind up being included in a single block.  Probably the more useful place to prioritise based on new validators is one level up so we pick the `MatchingDataAttestationGroup` to include in the block based on the reward it would earn.  That would be a function of the number of new validators and the inclusion distance.  Currently we're prioritising most recent first only.

**Comment by mbaxter** [comment]
> Probably the more useful place to prioritise based on new validators is one level up so we pick the `MatchingDataAttestationGroup` to include in the block based on the reward it would earn

Yeah, I agree.  I don't think this change ends up being very useful in and of itself and it does add additional processing.  I'm going to close this for now - we can look at a more comprehensive rework of the attestation prioritization when someone picks up #2192.

**Pull request closed without merging**